// zigbee-herdsman-converters/devices/smlight.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const m = __importStar(require("../lib/modernExtend"));
const definitions = [
    {
        zigbeeModel: ['SLZB-06M', 'SLZB-06', 'SLZB-06p7', 'SLZB-07', 'SLZB-0xp7', 'SLZB-07Mg24', 'SLZB-06Mg24'],
        model: 'SLZB-06M',
        vendor: 'SMLIGHT',
        description: 'Router',
        whiteLabel: [
            { vendor: 'SMLIGHT', model: 'SLZB-06', description: 'Router', fingerprint: [{ modelID: 'SLZB-06' }] },
            { vendor: 'SMLIGHT', model: 'SLZB-06p7', description: 'Router', fingerprint: [{ modelID: 'SLZB-06p7' }] },
            { vendor: 'SMLIGHT', model: 'SLZB-07', description: 'Router', fingerprint: [{ modelID: 'SLZB-07' }] },
            { vendor: 'SMLIGHT', model: 'SLZB-0xp7', description: 'Router', fingerprint: [{ modelID: 'SLZB-0xp7' }] },
            { vendor: 'SMLIGHT', model: 'SLZB-07Mg24', description: 'Router', fingerprint: [{ modelID: 'SLZB-07Mg24' }] },
            { vendor: 'SMLIGHT', model: 'SLZB-06Mg24', description: 'Router', fingerprint: [{ modelID: 'SLZB-06Mg24' }] },
        ],
        extend: [m.linkQuality({ reporting: true }), m.forcePowerSource({ powerSource: 'Mains (single phase)' })],
    },
];
exports.default = definitions;
module.exports = definitions;
